DataStructures/Graphs/MaxFlowDinic.java

```diff
@@ -0,0 +1,160 @@
+     * times for a single call of BFS.
+     * @param flow Current flow send by parent function call
+     * @param start To keep track of next edge to be explored; start[i] stores  count of edges explored from i.
+     * @param u Current vertex
+     * @param t Sink 
+     */
+    private double sendFlow(int u, double flow, int t, int start[])
+    {
+        // Sink reached
+        if (u == t)
+            return flow;
+
+        // Traverse all adjacent edges one -by - one.
+        for (  ; start[u] < adj[u].size(); start[u]++)
+        {
+            // Pick next edge from adjacency list of u
+            Edge e = adj[u].get(start[u]); 
+            if (level[e.v] == level[u]+1 && e.flow < e.cap)
+            {
+                // find minimum flow from u to t
+                double curr_flow = Math.min(flow, e.cap - e.flow);
+                double temp_flow = sendFlow(e.v, curr_flow, t, start);
+                // flow is greater than zero
+                if (temp_flow > 0)
+                {
+                    // add flow  to current edge
+                    e.flow += temp_flow;
+                    // subtract flow from reverse edge
+                    // of current edge
+                    adj[e.v].get(e.index).flow -= temp_flow;
+                    return temp_flow;
+                }
+            }
+        }
+        return 0;
+    }
+    /**
+     * @param s=Source vertex
+     * @param t=Sink Vertex
+     */
+    double DinicMaxflow(int s, int t)
+    {
+        // Corner case
+        if (s == t)
+            return -1;
+        double total = 0;  // Initialize result
+        // Augment the flow while there is path
+        // from source to sink
+        while (BFS(s, t))
+        {
+            // store how many edges are visited
+            // from V { 0 to V }
+            int start[] = new int[V+1];
+            // while flow is not zero in graph from S to D
+            while (true)
+            {
+                double flow = sendFlow(s, Integer.MAX_VALUE, t, start);
+                if(flow<=0)
+                    break;
+                total += flow;
+            }
+        }
+        return total;
+    }
+    public static void main(String args[])
+    {
+        MaxFlowDinic obj=new MaxFlowDinic(0);//Dummy object to access the non-static methods
+        obj.go();
+        
+    }
+    private void go()
+    {
+        MaxFlowDinic g=new MaxFlowDinic(6);
+        //Sample example. In order to use this class, user should create an object, populte the graph via addEdge() method and cal the DinicMaxFlow() method on it
+        g.addEdge(0, 1, 16 );
+        g.addEdge(0, 2, 13 );
+        g.addEdge(1, 2, 10 );
+        g.addEdge(1, 3, 12 );
+        g.addEdge(2, 1, 4 );
+        g.addEdge(2, 4, 14);
+        g.addEdge(3, 2, 9 );
+        g.addEdge(3, 5, 20 );
+        g.addEdge(4, 3, 7 );
+        g.addEdge(4, 5, 4);
+        System.out.println( "Maximum flow is: "+g.DinicMaxflow(0, 5));    
+    }
```

Comment:
If I compile this I get:
> Note: MaxFlowDinic.java uses unchecked or unsafe operations.
Note: Recompile with -Xlint:unchecked for details.

You must use type specifies ```<TYPE>``` by each collection.